### DISCUSSION TITLE
Combining complex commands to avoid temporary files?

### DISCUSSION THREAD

**Discussion by gingerbeardman** [ImageMagick/ImageMagick#4095] [Help]
Combining commands is not my strong point, so I thought I'd start a discussion here so I can learn how to go about it better.

I'm making a driving game for a handheld console called Playdate and my workflow involves rendering 3D to images and the processing them using IM to do some resizing, stitching, and image processing.

Read more: https://blog.gingerbeardman.com/2021/06/05/channelling-rgb-into-1bit/

My workflow for 35 vehicles takes 1 hour and generates 200K files taking up 1GB data. The rendering part of that workflow runs 2640 times (once per frame) per vehicle, and the IM part of that process only operates once per vehicle after the render is done.

Anyway! The renders are in red, green and blue (literally) and, after resizing and stitching, I split them into channels using IM and process each channel in a different way, then I composite the results back together. Finally I append a separate image to the bottom!

Final image: [car-table-38-38.png](https://user-images.githubusercontent.com/49612/129105825-c795f94d-2c93-4417-b879-6f6a4f2d3dc1.png)

My feeling is that I am creating temporary files when I shouldn't have to. 

How can I combine these steps to remove the need for temporary files?

```bash
SPRITE="$OUTPUT/_SheetsFrames/${vehicle}*.png"
SHEETSHADOW="$OUTPUT/_SheetsShadows/${vehicle}*.png"
FINAL="$OUTPUT/_Mono/${vehicle}${IMGTABLE}"

## TEMP FILES
GFILE="$OUTPUT/dailydriver-render.green.png"
DFILE="$OUTPUT/dailydriver-render.detail.png"
RFILE="$OUTPUT/dailydriver-render.result.png"

## TEMP IMAGE 1
magick $SPRITE -channel G -separate -threshold 15% $GFILE

## TEMP IMAGE 2
magick \( \( -size $IMGSIZE pattern:gray50 -alpha on \) \( $GFILE -transparent black -background white \) -compose Dst_In -composite \) $DFILE

## COMPOSITE CHANNELS to TEMP IMAGE 3
magick \
-page 0 \( $SPRITE -channel R -separate -threshold 15% -negate \) \
-page 0 \( $DFILE \) \
-page 0 \( $SPRITE -channel B -separate -threshold 15% -transparent black \) \
-flatten \
\( $SPRITE -channel A -separate \) -alpha off \
-compose copy_opacity \
-composite \
$RFILE

# APPEND SHADOW SHEET
magick $RFILE \( $SHEETSHADOW -fuzz 10% -transparent white \) -append +repage $FINAL
```

**Comment by snibgo** [comment]
The main principle of combining magick commands, avoiding temporary files, is to think of each one as an arithmetic calculation.

Suppose we have this arithmetic:
```
T = 3*4+5
R = 5+6+T*2
```
How do we calculate R directly, without bothering with T? We substitute T in the calculation of R:
```
R = 5+6+(3*4+5)*2
```
I have put the calculation of T in parentheses.

Combining magick commands is very similar, except that the use of parentheses is to ensure that image processing operations are performed only on certain images.

When you make a change, check that re-running the script on a single file makes the expected result. This makes debugging simple. If you make loads of change before checking, debugging is harder.

I usually start at the end, and work backwards. Your last two magick commands has: one command creates $RFILE, and the next uses %RFILE as the first input. So combining them is easy:
```
## COMPOSITE CHANNELS to TEMP IMAGE 3, and APPEND SHADOW SHEET
magick \
-page 0 \( $SPRITE -channel R -separate -threshold 15% -negate \) \
-page 0 \( $DFILE \) \
-page 0 \( $SPRITE -channel B -separate -threshold 15% -transparent black \) \
-flatten \
\( $SPRITE -channel A -separate \) -alpha off \
-compose copy_opacity \
-composite \
\( $SHEETSHADOW -fuzz 10% -transparent white \) -append +repage $FINAL
```
The command reads $SPRITE three times. This is inefficient. We can read it just once, and write it to a memory location "mpr", then read from memory when we need it:
```
## COMPOSITE CHANNELS to TEMP IMAGE 3, and APPEND SHADOW SHEET
magick \
-page 0 \( $SPRITE +write mpr:SPRT -channel R -separate -threshold 15% -negate \) \
-page 0 \( $DFILE \) \
-page 0 \( mpr:SPRT -channel B -separate -threshold 15% -transparent black \) \
-flatten \
\( mpr:SPRT -channel A -separate \) -alpha off \
-compose copy_opacity \
-composite \
\( $SHEETSHADOW -fuzz 10% -transparent white \) -append +repage $FINAL
```
I have called the mpr SPRT because that is close to the name SPRITE but isn't identical. If it was identical, I might get confused.

That is as far as I would go before checking the work. Proceed with the others, so you merge them all into a single magick command.

Watch out for settings. "-background white" is a setting, and so is "-compose Dst_In". I don't think any of your later commands rely on default settings for these, but I could be wrong.

**Reply by gingerbeardman** [reply]
Thank you! I will dig into this and report back.
